libparsec/crates/client/src/workspace/store/prevent_sync_pattern.rs

```diff
@@ -0,0 +1,87 @@
+use std::sync::Arc;
+
+use libparsec_platform_storage::workspace::{UpdateManifestData, WorkspaceStorage};
+use libparsec_types::prelude::*;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ApplyPreventSyncPatternError {
+    #[error("Decryption error")]
+    DecryptionError,
+    #[error(transparent)]
+    Internal(anyhow::Error),
+}
+
+pub(super) async fn ensure_prevent_sync_pattern_applied_to_wksp(
+    storage: &mut WorkspaceStorage,
+    device: Arc<LocalDevice>,
+    pattern: &Regex,
+) -> Result<(), ApplyPreventSyncPatternError> {
+    const PAGE_SIZE: u32 = 1000;
+
+    let fully_applied = storage
+        .set_prevent_sync_pattern(pattern)
+        .await
+        .map_err(ApplyPreventSyncPatternError::Internal)?;
+
+    if fully_applied {
+        return Ok(());
+    }
+
+    let mut offset = 0;
+    let mut updated_manifest = Vec::new();
+    loop {
+        let manifests = storage
+            .list_manifests(offset, PAGE_SIZE)
+            .await
+            .map_err(ApplyPreventSyncPatternError::Internal)?;
+
+        for encoded_manifest in &manifests {
+            // Only the prevent sync pattern could be applied to a folder type manifest.
+            // We assume that workspace manifest and folder manifest could both be deserialize as folder manifest.
+            if let Some(folder) = decode_folder(encoded_manifest, &device.local_symkey)? {
+                let new_folder =
+                    folder.apply_prevent_sync_pattern(Some(pattern), device.time_provider.now());
+                if new_folder != folder {
+                    updated_manifest.push(UpdateManifestData {
+                        entry_id: new_folder.base.id,
+                        encrypted: new_folder.dump_and_encrypt(&device.local_symkey),
+                        need_sync: new_folder.need_sync,
+                        base_version: new_folder.base.version,
+                    })
+                }
+            }
+        }
+
+        storage
+            .update_manifests(updated_manifest.drain(..))
+            .await
+            .map_err(ApplyPreventSyncPatternError::Internal)?;
+
+        // The manifests list is not filled to the page size,
+        // We consider that another call will result in an empty list, so we can stop here.
+        if manifests.len() < PAGE_SIZE as usize {
+            break;
+        }
+        updated_manifest.clear();
+        offset += PAGE_SIZE;
+    }
+    storage
+        .mark_prevent_sync_pattern_fully_applied(pattern)
+        .await
+        .map_err(ApplyPreventSyncPatternError::Internal)?;
+    Ok(())
+}
+
+fn decode_folder(
+    encoded: &[u8],
+    symkey: &SecretKey,
+) -> Result<Option<LocalFolderManifest>, ApplyPreventSyncPatternError> {
+    match LocalFolderManifest::decrypt_and_load(encoded, symkey) {
+        Ok(manifest) => Ok(Some(manifest)),
+        Err(DataError::Decryption) => Err(ApplyPreventSyncPatternError::DecryptionError),
```

Comment:
I don't think we need a dedicated error for this

see for instance what we do here:
https://github.com/Scille/parsec-cloud/blob/b0bbab24069b6fadd38699b8eeb8ebf166282051/libparsec/crates/client/src/workspace/store/cache.rs#L219-L222

Reply:
done

Reply:
would be better to have an `anyhow::anyhow!("Local database contains invalid data: {}", err),` to help in the error message instead of just `err.into()`